### PR TITLE
Support partitioned topics in the Reader

### LINKED ISSUE #3643: [client] Support seek at partitioned topic
**Is your feature request related to a problem? Please describe.**

Currently seeking is not supported on partitioned topics.

**Describe the solution you'd like**

It should be doable to support seek on partitioned topics.

Because:

1) seeking by earliest / latest can just seek on all partitions.
2) seeking by messageId is also doable. because messageId provides partition id.

**Describe alternatives you've considered**

No other alternative. We can just decide not to support it.

**Additional context**

It would be much easier for people to use if we support `seek` at partitioned topics.

### PR DESCRIPTION
**Pull request by 315157973**
Fixes #3643 #7265

### Motivation
Support partitioned topics in the Reader

### Modifications
class relationship：
PulsarClientImpl -> MultiTopicsReaderImpl -> MultiTopicsConsumerImpl -> ConsumerImpl

PulsarClientImpl support build MultiTopicsReader
MultiTopicsReader wraps one MultiTopicsConsumerImpl 
MultiTopicsConsumerImpl contains multiple consumerImpl

Make MultiTopicsConsumerImpl support seek by messageId

seek by time:
All consumerImpl seek by time, Reader can get all the messages returned by each partition

seek by message：
1）When seek by latest/earliest,all partitions seek by latest/earliest,Reader can get all the messages returned by each partition
2）When the messageId contains explicit partition information, it will only receive messages from this partition, and will not receive messages from other partitions.To avoid other partitions returning messages in inclusive mode, I make other consumer seek to latest and add a ignoredConsumersSet to ignore messages from other partitions .

### Verifying this change
unit tests:
TopicReaderTest
MultiTopicsReaderTest

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by sollecitom** [comment]
Any progress on this? :)

**Comment by sijie** [comment]
@sollecitom : @jiazhai @codelipenghui will review this pull request today

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java [hunk 1]
```diff
@@ -1272,4 +1389,35 @@ public Timeout getPartitionsAutoUpdateTimeout() {
...
+        }
+        return true;
+    }
+
+    public void tryAcknowledgeMessage(Message<T> msg) {
+        if (msg != null) {
+            BatchMessageIdImpl batchMessageId = new BatchMessageIdImpl(MessageIdImpl.convertToMessageIdImpl(msg.getMessageId()));
+            //Non-batching messages ack every time, batchMessage only need to ack the last one to avoid multi recycle
+            if (batchMessageId.getBatchIndex() < 0 || batchMessageId.getBatchSize() - 1 == batchMessageId.getBatchIndex()) {
+                acknowledgeCumulativeAsync(msg);
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:1419]: I think here also assume that all topics of the multiple topics consumer are under a partitioned topic?

**Review by codelipenghui — COMMENTED** [review]
The topics of the multiple topics consumer may belong to different partitioned topics, looks the PR assume topics of the multiple topics consumer belong to a partitioned topic, please double-check.

**Comment by 315157973** [comment]
There are still some details of this requirement that need to be discussed:
Suppose there is a partitioned topic that contains 3 partitions.
1) If a `MessageId` belongs to `partition-1`, when the reader sets startMessageId, how should the messages of the other two partitions be processed? earliest, latest...?
2) If the user first seeks (earliest) and then seeks a certain `MessageId`, how should the messages of the other two partitions be processed? Forbid the other 2 partitions to return messages? Or continue to let the other 2 partitions return messages?

Thanks for @codelipenghui 's suggestion:
In this PR, for MultiTopicReader, we only support seek `earliest` and `latest` , and do not support seek with a specific `MessageId`

**Comment by slinstaedt** [comment]
Seeking to a specific MessageId seems to be mandatory though. How should one resume reading otherwise?

**Review by codelipenghui — APPROVED** [review]
The change looks good to me.

**Comment by codelipenghui** [comment]
@slinstaedt We will support the reader to seek to the earliest position or the latest position first. Seek to a specific position is confusing in the partitioned reader, we can try to support later since there is not a good idea yet.

**Comment by codelipenghui** [comment]
@sijie @jiazhai Please help review this PR, thanks.

**Review by sijie — APPROVED** [review]

**Comment by sijie** [comment]
@315157973 Can you rebase it to the latest master?

**Pull request merged**